runtime/sema/check_variable_declaration.go

```diff
@@ -234,3 +236,119 @@ func (checker *Checker) elaborateNestedResourceMoveExpression(expression ast.Exp
+		return
+	}
+}
+
+func (checker *Checker) recordReference(targetVarName string, expr ast.Expression) {
+	referencedVar := checker.referencedVariable(expr)
+
+	if referencedVar != nil &&
+		referencedVar.Type.IsResourceType() {
+		checker.references.Set(targetVarName, referencedVar)
+	}
+}
+
+// referencedVariable return the referenced variable
+func (checker *Checker) referencedVariable(expr ast.Expression) *Variable {
+	refExpression := checker.referenceExpression(expr)
+
+	var variableRefExpr *ast.Identifier
+
+	switch refExpression := refExpression.(type) {
+	case *ast.ReferenceExpression:
+		// If it is a reference expression, then find the root variable.
+		// i.e: root variable of `&a.b.c as &T` is `a`.
+		// This is because nested resources cannot be moved.
+		// If `c` needs to be moved, then `a` has to be moved.
+		// So tracking `a` is sufficient.
+		variableRefExpr = checker.variableReferenceExpression(refExpression.Expression)
+	case *ast.IdentifierExpression:
+		variableRefExpr = &refExpression.Identifier
+	default:
+		return nil
+	}
+
+	if variableRefExpr == nil {
+		return nil
+	}
+
+	referencedVariableName := variableRefExpr.Identifier
+
+	for {
+		// If the referenced variable is again a reference,
+		// then find the variable of the root of the reference chain.
+		// e.g::
+		//     ref1 = &v
+		//     ref2 = &ref1[0]
+		//     ref2.field = 3
+		//
+		// Here, `ref2` refers to `ref1`, which refers to `v`.
+		// So `ref2` is actually referring to `v`
+
+		referencedRef := checker.references.Find(referencedVariableName)
+		if referencedRef == nil {
+			break
+		}
+
+		referencedVariableName = referencedRef.Identifier
+	}
+
+	return checker.valueActivations.Find(referencedVariableName)
+}
+
+// referenceExpression returns the expression that return a reference.
+// There could be two types of expressions that can result in a reference:
+//  1. Expressions that create a new reference.
+//     (i.e: reference-expression)
+//  2. Expressions that returns an existing reference.
+//     (i.e: identifier-expression/member-expression/index-expression having a reference type)
+//
+// However, it is  not currently possible to track member-expressions/index-expression.
+// So this method would only return either a `reference-expression` or an `identifier-expression`.
+//
+// The expression could also be hidden inside some other expression.
+// e.g(1): `&v as &T` is a casting expression, but has a hidden reference expression.
+// e.g(2): `(&v as &T?)!
+func (checker *Checker) referenceExpression(expr ast.Expression) ast.Expression {
+	switch expr := expr.(type) {
+	case *ast.ReferenceExpression:
+		return expr
+	case *ast.ForceExpression:
+		return checker.referenceExpression(expr.Expression)
+	case *ast.CastingExpression:
+		return checker.referenceExpression(expr.Expression)
+	case *ast.BinaryExpression:
+		if expr.Operation != ast.OperationNilCoalesce {
+			return nil
+		}
+		return checker.referenceExpression(expr.Left)
```

Comment:
Why do we return the left expression for nil-coalescing? Is it a "best guess"?

Reply:
Resources re not allowed in rhs of the nil-coalescing expressions. So only need to visit the lhs.
I'll add a comment mentioning this.

Reply:
ah actually, reference is still allowed. good catch!

Reply:
Added: https://github.com/onflow/cadence/pull/2037/commits/c6dff546a8d9a6bc35e635fa3aae09b386c8a5db